File: python/tinymq.py
from bottle import run, debug, abort, request, ServerAdapter, response, get, post
import redis
import json
import gevent
from gevent.queue import Queue
from Queue import Empty

from gevent import monkey
from collections import defaultdict
monkey.patch_all()

QUEUESET = 'QUEUESET'
UUID_SUFFIX = ':UUID'
QUEUE_SUFFIX = ':queue'

redis_cli = redis.Redis()
_queue_presence = defaultdict(lambda :Queue())

def queue_hub():
    while True:
        queues = list(redis_cli.smembers(QUEUESET))
        queues = map(lambda x : "%s%s" % (x, QUEUE_SUFFIX), queues)
        if len(queues) > 0:
            res = redis_cli.brpop(queues)
            if res is not None: 
                _queue_presence[res[0]].put(res[1])

gevent.spawn(queue_hub)

@get('/')
def all_queues():
    qs = redis_cli.smembers(QUEUESET) 
    if qs is None: abort(404, "No queues created")
    l = ["/q/%s" % q for q in qs]
    return json.dumps(l)

@get('/c/:q')
def queue(q = None):
    if q is None: abort(404, "No queue given")
    qn = q + QUEUE_SUFFIX
    response.set_header("Transfer-Encoding", "chunked")
    response.content_type = 'application/json; charset=UTF-8'
    redis_cli.sadd(QUEUESET, q)
    v = b = None
    yield " \n\n" # clean the pipes
    while(True):
        try:
            b = _queue_presence[qn].get(block=True, timeout=60)
            if b is not None: v = redis_cli.get(b)
        except Empty, e:
            pass

        if v is not None: 
            yield "%s\n" % json.dumps({"value": v, "key": b})
            v = None

@get('/q/:q')
def queue(q = None):
    if q is None: abort(404, "No queue given")
    redis_cli.sadd(QUEUESET, q)
    q = q + QUEUE_SUFFIX

    if "soft" in request.GET.keys(): b = redis_cli.lindex(q, -1)
    else: b = redis_cli.rpop(q)

    if b is None: abort(404, "Empty queue")
    v = redis_cli.get(b)

    if v is None: abort(404, "Empty value/no val")
    return json.dumps({"value": v, "key": b})

@post('/q/:q')
def queue(q = None):
    if q is None: abort(404, "No queue given")
    qkey = q + QUEUE_SUFFIX

    value = request.POST['value']
    if value is None: abort(401, "No value given")

    uuid = redis_cli.incr(q + UUID_SUFFIX)
    redis_cli.sadd(QUEUESET, q)

    lkey = "%s:%d" % (q, uuid)
    redis_cli.set(lkey, value)
    redis_cli.lpush(qkey, lkey)
    return json.dumps({"ok": lkey})

class GEventServerAdapter(ServerAdapter):
    def run(self, handler):
        from gevent import monkey
        monkey.patch_socket()
        from gevent.wsgi import WSGIServer
        WSGIServer((self.host, self.port), handler).serve_forever()

#debug(True)
run(host='localhost', port=8081, server=GEventServerAdapter)

File: ruby/tinymq.rb
require 'rubygems'
require "bundler/setup"
require 'sinatra'
require 'redis'
require 'json'

QUEUESET = 'QUEUESET'   # queue index
UUID_SUFFIX = ':UUID'   # queue unique id
QUEUE_SUFFIX = ':queue' # suffix to identify each queue's LIST

set :server, :thin

reds = Redis.new
_queue_presence = Hash.new { |h,k| h[k] = Array.new }

Thread.new do
  while true do
    queues = reds.smembers QUEUESET
    queues.map! do |q| "#{q}#{QUEUE_SUFFIX}" end
    if queues.size > 0 then
      q, k= reds.brpop *queues, 5
      v = reds.get k
      _queue_presence[q].each do |out| 
        b = {"value"=>v, "key"=>k} 
        out << b.to_json << "\n"
      end
    end
  end
end

get '/q' do
    b = reds.smembers QUEUESET
    throw :halt, [404, 'Not found (empty queueset)'] if b == nil
    b.map! do |q| q = '/q/'+q end
    b.to_json
end

get '/q/:queue' do |queue|
  soft = params['soft'] # soft = true doesn't rpop values
  throw :halt, [404, 'Not found'] if queue == nil
  queue = queue + QUEUE_SUFFIX
  if soft != nil
    b = reds.lindex queue, -1
  else
    b = reds.rpop queue 
  end
  throw :halt, [404, 'Not found (empty queue)'] if b == nil
  v = reds.get b
  r = {"value"=>v, "key"=>b} 
  throw :halt, [200, r.to_json] unless v == nil 
  'empty value'
end

post '/q/*' do |queue|
  value = params['value'].to_s
  throw :halt, [404, "Not found"] if queue == nil
  q1 = queue + QUEUE_SUFFIX
  uuid = reds.incr queue + UUID_SUFFIX 
  reds.sadd QUEUESET, q1
  lkey = queue + ':' + uuid.to_s
  reds.set lkey, value
  reds.lpush q1, lkey
  body '{ok, ' + lkey + '}'
end

get '/c/*' do |queue|
  queue = queue + QUEUE_SUFFIX
  stream(keep_open=true) do |out|
      _queue_presence[queue] << out
  end
end
